File: 10 October/10-20/tip_calculator.py
def calculate_tips(meal_price, custom_tip):
    import re

    meal_price_dict = re.match(r"\$(?P<meal_price_value>\d+\.\d+)", meal_price).groupdict()
    custom_tip_dict = re.match(r"(?P<custom_tip_value>\d+)%", custom_tip).groupdict()

    tip15 = f"${float(meal_price_dict['meal_price_value']) * 15 / 100 :.2f}"
    tip20 = f"${float(meal_price_dict['meal_price_value']) * 20 / 100 :.2f}"
    tip_custom = f"${float(meal_price_dict['meal_price_value']) * float(custom_tip_dict['custom_tip_value']) / 100 :.2f}"

    return [ tip15, tip20, tip_custom ]

print(calculate_tips("$10.00", "25%"))
print(calculate_tips("$89.67", "26%"))
print(calculate_tips("$19.85", "9%"))

# --- TEST SUITE ---

tests_text = '''
1. calculate_tips("$10.00", "25%") should return ["$1.50", "$2.00", "$2.50"].
2. calculate_tips("$89.67", "26%") should return ["$13.45", "$17.93", "$23.31"].
3. calculate_tips("$19.85", "9%") should return ["$2.98", "$3.97", "$1.79"].
'''

import re

tests_regex = re.compile(r"(?P<number>\d+)\.\s(?P<function_call>.+) should return (?P<output>.+?)\.?$", re.MULTILINE)
test_data = [item.groupdict() for item in re.finditer(tests_regex, tests_text)]

def run_tests(test_data):
    print("--------------------------")
    print("🧪 Starting Verification...")
    print("--------------------------")

    all_passed = True
    fail_count = 0

    for test in test_data:
        function_call_output = eval(test['function_call'])
        test_output = eval(test['output'])
        
        if function_call_output == test_output:
            print(f"{test['number']}.✅ PASS - Function Call:\n{test['function_call']}")
        else:
            print(f"{test['number']}.❌ FAIL - Function Call:\n{test['function_call']}\nExpected:\n{test_output}\nGot:\n{function_call_output}")
            all_passed = False
            fail_count += 1

    print("----------------------------");

    if (all_passed):
        print("🎉 SUCCESS: All tests PASSED.");
    else:
        print(f"⚠️ WARNING: {fail_count}/{len(test_data)} tests FAILED.");

run_tests(test_data)

File: 10 October/10-20/tipCalculator.js
function calculateTips(mealPrice, customTip) {
  const { mealPriceValue } = /\$(?<mealPricevalue>\d+\.\d+)/.exec(mealPrice).groups;
  const { customTipValue } = /(?<customTipValue>\d+)%/.exec(customTip).groups;
  
  const tip15 = `$${parseFloat(mealPriceValue * 15 / 100).toFixed(2)}`;
  const tip20 = `$${parseFloat(mealPriceValue * 20 / 100).toFixed(2)}`;
  const tipCustom = `$${parseFloat(mealPriceValue * customTipValue / 100).toFixed(2)}`;

  return [ tip15, tip20, tipCustom ];
}

console.log(calculateTips("$10.00", "25%"));
console.log(calculateTips("$89.67", "26%"));
console.log(calculateTips("$19.85", "9%"));

// --- TEST SUITE ---

const testsText = `
1. calculateTips("$10.00", "25%") should return ["$1.50", "$2.00", "$2.50"].
2. calculateTips("$89.67", "26%") should return ["$13.45", "$17.93", "$23.31"].
3. calculateTips("$19.85", "9%") should return ["$2.98", "$3.97", "$1.79"].
`;

const testsRegex = /(?<number>\d+)\.\s(?<functionCall>.+) should return (?<output>.+?)\.?$/gm;
const testData = [...testsText.matchAll(testsRegex).map(match => match.groups)];

function runTests(testData) {
  console.log("--------------------------");
  console.log("🧪Starting Verification...");
  console.log("--------------------------");

  let allPassed = true;
  let failCount = 0;
  
  testData.forEach(test => {
    const functionCallOutput = eval(test.functionCall);
    const testOutput = eval(test.output);
    let comparison;

    if (Array.isArray(testOutput))
      comparison = arraysEqual(functionCallOutput, testOutput);
    else
      comparison = functionCallOutput === testOutput;

    if (comparison) {
      console.log(`${test.number}.✅PASS - Function Call: ${test.functionCall}`);
    }
    else {
      console.log(`${test.number}.❌FAIL - Function Call: ${test.functionCall}\nExpected: ${testOutput}\nGot: ${functionCallOutput}`);
      allPassed = false;
      failCount ++;
    }
  })

  console.log("----------------------------");

  if (allPassed)
    console.log("🎉SUCCESS: All tests PASSED.");
  else
    console.log(`⚠️WARNING: ${failCount}/${testData.length} tests FAILED.`);
}

function arraysEqual(a, b) {
  if (a.length !== b.length)
    return false;

  for (const index in a) {
    if (Array.isArray(a[index])) {
      if (Array.isArray(b[index])) {
        if (!arraysEqual(a[index], b[index]))
          return false;
      }
      else 
        return false;
    }

    else if (a[index] !== b[index])
      return false;
  }

  return true;
}

runTests(testData);